### PR TITLE
Refactored the fsScanner into a separate package

### PR DESCRIPTION
**Pull request by Jille**
Had to introduce a common package to share some types/funcs

The new version does basically the same but is split up in many more functions which I'm told makes things more readable.
It also allows for multiple listeners which will make multi-master easier one day.
It also allows for multiple shares to be implemented in a cleaner way.

### CODE REVIEW

**Review by dsprenkels — APPROVED** [review]
All my suggestions are included in 32e0275e8348559de8c1f3cba4354946906b533b.

**Comment by dsprenkels** [comment]
Btw. Could it be possible to merge https://github.com/Jille/rufs/pull/16 into this PR? I feel this PR is going to break that one.

File: Gopkg.lock [hunk 1]
```diff
@@ -13,6 +13,17 @@
   pruneopts = ""
   revision = "371fbbdaa8987b715bdd21d6adc4c9b20155f748"
 
+[[projects]]
```
Comment (Jille) [inline Gopkg.lock]: I don't think this should be added?

Comment (dsprenkels) [reply Gopkg.lock]: I dunno? `dep ensure` generated this file for me.

**Review by dsprenkels — CHANGES_REQUESTED** [review]
I seem to be getting a data race with these changes:

```
dec 28 15:43:06 roku systemd[1]: Stopping RUFS katarastorage export...
dec 28 15:43:06 roku rufs[5003]: ==================
dec 28 15:43:06 roku rufs[5003]: WARNING: DATA RACE
dec 28 15:43:06 roku rufs[5003]: Write at 0x00c42028ec60 by goroutine 15:
dec 28 15:43:06 roku rufs[5003]:   runtime.closechan()
dec 28 15:43:06 roku rufs[5003]:       /usr/lib/go-1.10/src/runtime/chan.go:325 +0x0
dec 28 15:43:06 roku rufs[5003]:   main.(*Server).Run()
dec 28 15:43:06 roku rufs[5003]:       /home/daan/go/src/github.com/dsprenkels/rufs/server.go:141 +0x137
dec 28 15:43:06 roku rufs[5003]:   main.main.func1()
dec 28 15:43:06 roku rufs[5003]:       /home/daan/go/src/github.com/dsprenkels/rufs/main.go:111 +0x48
dec 28 15:43:06 roku rufs[5003]: Previous read at 0x00c42028ec60 by goroutine 22:
dec 28 15:43:06 roku rufs[5003]:   runtime.chansend()
dec 28 15:43:06 roku rufs[5003]:       /usr/lib/go-1.10/src/runtime/chan.go:140 +0x0
dec 28 15:43:06 roku rufs[5003]:   main.(*Server).SetFile-fm()
dec 28 15:43:06 roku rufs[5003]:       /home/daan/go/src/github.com/dsprenkels/rufs/server.go:171 +0xa4
dec 28 15:43:06 roku rufs[5003]:   github.com/dsprenkels/rufs/vendor/github.com/Jille/rufs/filescanner.(*Scanner).set()
dec 28 15:43:06 roku rufs[5003]:       /home/daan/go/src/github.com/dsprenkels/rufs/vendor/github.com/Jille/rufs/filescanner/filescanner.go:100 +0x527
dec 28 15:43:06 roku rufs[5003]:   github.com/dsprenkels/rufs/vendor/github.com/Jille/rufs/filescanner.(*Scanner).normalPass.func1()
dec 28 15:43:06 roku rufs[5003]:       /home/daan/go/src/github.com/dsprenkels/rufs/vendor/github.com/Jille/rufs/filescanner/filescanner.go:176 +0x4e5
dec 28 15:43:06 roku rufs[5003]:   github.com/dsprenkels/rufs/vendor/github.com/Jille/rufs/filescanner.scanFilesystemAndReportMissing.func1()
dec 28 15:43:06 roku rufs[5003]:       /home/daan/go/src/github.com/dsprenkels/rufs/vendor/github.com/Jille/rufs/filescanner/filescanner.go:223 +0x9f
dec 28 15:43:06 roku rufs[5003]:   github.com/dsprenkels/rufs/vendor/github.com/Jille/rufs/filescanner.scanFilesystem.func1()
dec 28 15:43:06 roku rufs[5003]:       /home/daan/go/src/github.com/dsprenkels/rufs/vendor/github.com/Jille/rufs/filescanner/filescanner.go:247 +0x3c8
dec 28 15:43:06 roku rufs[5003]:   path/filepath.walk()
dec 28 15:43:06 roku rufs[5003]:       /usr/lib/go-1.10/src/path/filepath/path.go:357 +0x501
dec 28 15:43:06 roku rufs[5003]:   path/filepath.walk()
dec 28 15:43:06 roku rufs[5003]:       /usr/lib/go-1.10/src/path/filepath/path.go:381 +0x3b6
dec 28 15:43:06 roku rufs[5003]:   path/filepath.walk()
dec 28 15:43:06 roku rufs[5003]:       /usr/lib/go-1.10/src/path/filepath/path.go:381 +0x3b6
dec 28 15:43:06 roku rufs[5003]:   path/filepath.walk()
dec 28 15:43:06 roku rufs[5003]:       /usr/lib/go-1.10/src/path/filepath/path.go:381 +0x3b6
dec 28 15:43:06 roku rufs[5003]:   path/filepath.walk()
dec 28 15:43:06 roku rufs[5003]:       /usr/lib/go-1.10/src/path/filepath/path.go:381 +0x3b6
dec 28 15:43:06 roku rufs[5003]:   path/filepath.Walk()
dec 28 15:43:06 roku rufs[5003]:       /usr/lib/go-1.10/src/path/filepath/path.go:403 +0x14a
dec 28 15:43:06 roku rufs[5003]:   github.com/dsprenkels/rufs/vendor/github.com/Jille/rufs/filescanner.scanFilesystem()
dec 28 15:43:06 roku rufs[5003]:       /home/daan/go/src/github.com/dsprenkels/rufs/vendor/github.com/Jille/rufs/filescanner/filescanner.go:229 +0xde
dec 28 15:43:06 roku rufs[5003]:   github.com/dsprenkels/rufs/vendor/github.com/Jille/rufs/filescanner.scanFilesystemAndReportMissing()
dec 28 15:43:06 roku rufs[5003]:       /home/daan/go/src/github.com/dsprenkels/rufs/vendor/github.com/Jille/rufs/filescanner/filescanner.go:221 +0xd5
dec 28 15:43:06 roku rufs[5003]:   github.com/dsprenkels/rufs/vendor/github.com/Jille/rufs/filescanner.(*Scanner).normalPass()
dec 28 15:43:06 roku rufs[5003]:       /home/daan/go/src/github.com/dsprenkels/rufs/vendor/github.com/Jille/rufs/filescanner/filescanner.go:163 +0x104
dec 28 15:43:06 roku rufs[5003]:   github.com/dsprenkels/rufs/vendor/github.com/Jille/rufs/filescanner.(*Scanner).normalPassWithIntermediateAutoSaving()
dec 28 15:43:06 roku rufs[5003]:       /home/daan/go/src/github.com/dsprenkels/rufs/vendor/github.com/Jille/rufs/filescanner/filescanner.go:157 +0xcb
dec 28 15:43:06 roku rufs[5003]:   github.com/dsprenkels/rufs/vendor/github.com/Jille/rufs/filescanner.(*Scanner).Run()
dec 28 15:43:06 roku rufs[5003]:       /home/daan/go/src/github.com/dsprenkels/rufs/vendor/github.com/Jille/rufs/filescanner/filescanner.go:107 +0x7b
dec 28 15:43:06 roku rufs[5003]: Goroutine 15 (running) created at:
dec 28 15:43:06 roku rufs[5003]:   main.main()
dec 28 15:43:06 roku rufs[5003]:       /home/daan/go/src/github.com/dsprenkels/rufs/main.go:110 +0x459
dec 28 15:43:06 roku rufs[5003]: Goroutine 22 (running) created at:
dec 28 15:43:06 roku rufs[5003]:   main.(*Server).Run()
dec 28 15:43:06 roku rufs[5003]:       /home/daan/go/src/github.com/dsprenkels/rufs/server.go:138 +0x707
dec 28 15:43:06 roku rufs[5003]:   main.main.func1()
dec 28 15:43:06 roku rufs[5003]:       /home/daan/go/src/github.com/dsprenkels/rufs/main.go:111 +0x48
dec 28 15:43:06 roku rufs[5003]: ==================
```

File: Gopkg.lock (hunk 1 above)
Comment (Jille) [reply Gopkg.lock]: Where you at the autoip branch at the time?

**Comment by Jille** [comment]
I fixed the data race at https://github.com/Jille/rufs/pull/21/files#diff-34c6b408d72845d076d47126c29948d1R179

**Review by dsprenkels — APPROVED** [review]

**Comment by sgielen** [comment]
As of 640c401637929d7adf2daba27ece5c7465af09d8 the software version is displayed in logs, UI and API calls to discovery. #24 covers metrics.

**Pull request merged**